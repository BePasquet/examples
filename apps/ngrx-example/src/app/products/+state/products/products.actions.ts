import { createAction, props } from '@ngrx/store';
import { Product, ProductFilter } from '../../../data';

export const getProducts = createAction(
  '[Products] Get Products',
  props<{ payload: ProductFilter }>()
);

export const getProductsSuccess = createAction(
  '[Products] Get Products Success',
  props<{ payload: Product[] }>()
);

export const getProductsFail = createAction(
  '[Products] Get Products Fail',
  props<{ payload: string }>()
);

export const createProduct = createAction(
  '[Products] Create Product',
  props<{ payload: Omit<Product, 'id'> }>()
);

export const createProductSuccess = createAction(
  '[Products] Create Product Success',
  props<{ payload: Product }>()
);

export const createProductFail = createAction(
  '[Products] Create Product Fail',
  props<{ payload: string }>()
);

export const updateProduct = createAction(
  '[Products] Update Product',
  props<{ payload: Product }>()
);

export const updateProductSuccess = createAction(
  '[Products] Update Product Success',
  props<{ payload: Product }>()
);

export const updateProductFail = createAction(
  '[Products] Update Product Fail',
  props<{ payload: string }>()
);

export const deleteProduct = createAction(
  '[Products] Delete Product',
  props<{ payload: { productId: string } }>()
);

export const deleteProductSuccess = createAction(
  '[Products] Delete Product Success',
  props<{ payload: { productId: string } }>()
);

export const deleteProductFail = createAction(
  '[Products] Delete Product Fail',
  props<{ payload: string }>()
);

export const setSelectedProductId = createAction(
  '[Products] Set Selected Product Id',
  props<{ payload: { productId: string } }>()
);

export type ProductActions = ReturnType<
  | typeof getProducts
  | typeof getProductsSuccess
  | typeof getProductsFail
  | typeof createProduct
  | typeof createProductSuccess
  | typeof createProductFail
  | typeof updateProduct
  | typeof updateProductSuccess
  | typeof updateProductFail
  | typeof deleteProduct
  | typeof deleteProductSuccess
  | typeof deleteProductFail
  | typeof setSelectedProductId
>;
