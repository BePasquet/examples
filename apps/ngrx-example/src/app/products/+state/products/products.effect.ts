import { Injectable } from '@angular/core';
import { Actions, createEffect, ofType } from '@ngrx/effects';
import { of } from 'rxjs';
import {
  catchError,
  concatMap,
  exhaustMap,
  map,
  mergeMap,
  switchMap,
} from 'rxjs/operators';
import { ProductsService } from '../../services/products.service';
import {
  createProduct,
  createProductFail,
  createProductSuccess,
  deleteProduct,
  deleteProductFail,
  deleteProductSuccess,
  getProducts,
  getProductsFail,
  getProductsSuccess,
  updateProduct,
  updateProductFail,
  updateProductSuccess,
} from './products.actions';

@Injectable()
export class ProductsEffect {
  getProducts$ = createEffect(() =>
    this.actions$.pipe(
      ofType(getProducts),
      switchMap(({ payload }) =>
        this.productsService.getProducts(payload).pipe(
          map((response) => getProductsSuccess({ payload: response })),
          catchError(({ error }) => of(getProductsFail({ payload: error })))
        )
      )
    )
  );

  createProduct$ = createEffect(() =>
    this.actions$.pipe(
      ofType(createProduct),
      exhaustMap(({ payload }) =>
        this.productsService.createProduct(payload).pipe(
          map((response) => createProductSuccess({ payload: response })),
          catchError(({ error }) => of(createProductFail({ payload: error })))
        )
      )
    )
  );

  updateProduct$ = createEffect(() =>
    this.actions$.pipe(
      ofType(updateProduct),
      concatMap(({ payload }) =>
        this.productsService.updateProduct(payload).pipe(
          map(() => updateProductSuccess({ payload })),
          catchError(({ error }) => of(updateProductFail({ payload: error })))
        )
      )
    )
  );

  deleteProduct$ = createEffect(() =>
    this.actions$.pipe(
      ofType(deleteProduct),
      mergeMap(({ payload: { productId } }) =>
        this.productsService.deleteProduct(productId).pipe(
          map(() => deleteProductSuccess({ payload: { productId } })),
          catchError(({ error }) => of(deleteProductFail({ payload: error })))
        )
      )
    )
  );

  constructor(
    private readonly actions$: Actions,
    private readonly productsService: ProductsService
  ) {}
}
