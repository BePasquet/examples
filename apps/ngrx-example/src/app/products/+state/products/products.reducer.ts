import { createEntityAdapter, EntityState } from '@ngrx/entity';
import { createReducer, on } from '@ngrx/store';
import { Product } from '../../../data';
import {
  createProduct,
  createProductFail,
  createProductSuccess,
  deleteProduct,
  deleteProductFail,
  deleteProductSuccess,
  getProducts,
  getProductsFail,
  getProductsSuccess,
  updateProduct,
  updateProductFail,
  updateProductSuccess,
} from './products.actions';

export const PRODUCTS_STATE_KEY = 'products';

export interface ProductsState extends EntityState<Product> {
  total: number;
  loading: boolean;
  loaded: boolean;
  error: string;
}

export interface ProductsPartialState {
  [PRODUCTS_STATE_KEY]: ProductsState;
}

export const productsAdapter = createEntityAdapter<Product>();

export const productsInitialState = productsAdapter.getInitialState({
  total: 0,
  loading: false,
  loaded: false,
  error: '',
});

export const productsReducer = createReducer(
  productsInitialState,
  on(getProducts, (state) =>
    productsAdapter.removeAll({
      ...state,
      loading: true,
      loaded: false,
      error: '',
      total: 0,
    })
  ),
  on(getProductsSuccess, (state, { payload: { results, total } }) =>
    productsAdapter.setAll(results, {
      ...state,
      loading: false,
      loaded: true,
      error: '',
      total,
    })
  ),
  on(getProductsFail, (state, { payload }) => ({
    ...state,
    loading: false,
    loaded: false,
    error: payload,
  })),
  on(createProduct, updateProduct, deleteProduct, (state) => ({
    ...state,
    loading: true,
    error: '',
  })),
  on(
    createProductFail,
    updateProductFail,
    deleteProductFail,
    (state, { payload }) => ({ ...state, loading: false, error: payload })
  ),
  on(createProductSuccess, (state, { payload }) =>
    productsAdapter.addOne(payload, {
      ...state,
      loading: false,
      error: '',
      total: state.total + 1,
    })
  ),
  on(updateProductSuccess, (state, { payload }) =>
    productsAdapter.updateOne(
      { id: payload.id, changes: { ...payload } },
      { ...state, loading: false, error: '' }
    )
  ),
  on(deleteProductSuccess, (state, { payload: { productId } }) =>
    productsAdapter.removeOne(productId, {
      ...state,
      loading: false,
      error: '',
      total: state.total - 1,
    })
  )
);
