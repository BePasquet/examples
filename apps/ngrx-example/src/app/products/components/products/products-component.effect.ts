import { Injectable } from '@angular/core';
import { Actions, createEffect, ofType } from '@ngrx/effects';
import { map } from 'rxjs/operators';
import { getProducts } from '../../+state';
import { searchProducts } from './products.component';

@Injectable()
export class ProductsComponentEffect {
  searchProducts$ = createEffect(() =>
    this.actions$.pipe(
      ofType(searchProducts),
      map(({ payload }) => getProducts({ payload }))
    )
  );

  constructor(private readonly actions$: Actions) {}
}
