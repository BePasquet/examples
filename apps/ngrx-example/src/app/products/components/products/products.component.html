<div>
  <h1>Products</h1>

  <div>
    <mat-form-field>
      <mat-label>Search</mat-label>
      <input
        #searchInputRef
        (input)="searchInput$.next(searchInputRef.value)"
        matInput
      />
      <mat-icon matSuffix>search</mat-icon>
    </mat-form-field>

    <button
      mat-raised-button
      color="primary"
      (click)="createUpdateClick$.next({ productId: '' })"
    >
      <mat-icon>add</mat-icon>
    </button>

    <button
      mat-raised-button
      color="primary"
      (click)="deleteClick$.next({ productId: '' })"
    >
      <mat-icon>remove</mat-icon>
    </button>
  </div>

  <div>
    {{ state$ | async | json }}
  </div>
</div>
