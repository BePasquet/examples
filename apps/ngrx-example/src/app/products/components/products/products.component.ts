import { Component, OnDestroy } from '@angular/core';
import {
  createAction,
  createSelector,
  props,
  select,
  Store,
} from '@ngrx/store';
import { merge, Subject, Subscription } from 'rxjs';
import {
  concatMap,
  debounceTime,
  distinctUntilChanged,
  map,
  startWith,
  tap,
} from 'rxjs/operators';
import {
  ProductsPartialState,
  selectProducts,
  selectProductsError,
  selectProductsLoaded,
  selectProductsLoading,
  setSelectedProductId,
} from '../../+state';
import { ProductFilter } from '../../../data';

// Actions
export const searchProducts = createAction(
  '[Products Component] Search Products',
  props<{ payload: ProductFilter }>()
);

export const openProductDialog = createAction(
  '[Products Component] Open Product Dialog'
);

export const openConfirmationDialog = createAction(
  '[Products Component] Open Confirmation Dialog',
  props<{ payload: { productId: string } }>()
);

// Selectors
export const selectProductsVM = createSelector(
  selectProducts,
  selectProductsLoading,
  selectProductsLoaded,
  selectProductsError,
  (products, loading, loaded, error) => ({ products, loading, loaded, error })
);

@Component({
  selector: 'products',
  templateUrl: './products.component.html',
  styleUrls: ['./products.component.scss'],
})
export class ProductsComponent implements OnDestroy {
  readonly state$ = this.store.pipe(select(selectProductsVM));

  // Controls
  readonly searchInput$ = new Subject<string>();

  readonly createUpdateClick$ = new Subject<{ productId: string }>();

  readonly deleteClick$ = new Subject<{ productId: string }>();

  private readonly searchProducts$ = this.searchInput$.pipe(
    startWith(''),
    debounceTime(350),
    distinctUntilChanged(),
    map((name) => searchProducts({ payload: { name } }))
  );

  private readonly showProductDialog$ = this.createUpdateClick$.pipe(
    concatMap(({ productId }) => [
      setSelectedProductId({ payload: { productId } }),
      openProductDialog(),
    ])
  );

  private readonly showConfirmationDialog$ = this.deleteClick$.pipe(
    map(({ productId }) => openConfirmationDialog({ payload: { productId } }))
  );

  private readonly componentActions$ = merge(
    this.searchProducts$,
    this.showProductDialog$,
    this.showConfirmationDialog$
  ).pipe(tap((action) => this.store.dispatch(action)));

  private readonly subscriptions = new Subscription();

  constructor(private readonly store: Store<ProductsPartialState>) {
    this.subscriptions.add(this.componentActions$.subscribe());
  }

  ngOnDestroy(): void {
    this.subscriptions.unsubscribe();
  }
}
