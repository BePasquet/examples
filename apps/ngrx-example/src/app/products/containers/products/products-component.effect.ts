import { Injectable } from '@angular/core';
import { MatDialog } from '@angular/material/dialog';
import { Actions, createEffect, ofType } from '@ngrx/effects';
import { select, Store } from '@ngrx/store';
import { concatMap, exhaustMap, map, withLatestFrom } from 'rxjs/operators';
import {
  createProductFail,
  createProductSuccess,
  deleteProduct,
  deleteProductFail,
  deleteProductSuccess,
  getProducts,
  ProductsPartialState,
  selectProductEntities,
  updateProductFail,
  updateProductSuccess,
} from '../../+state';
import { closeDialogs, showMessage } from '../../../layout';
import { ConfirmationDialogComponent } from '../../../shared/components/confirmation-dialog.component';
import { ProductFormComponent } from '../../components/product-form/product-form.component';
import {
  acceptDeleteConfirmationDialog,
  cancelDeleteConfirmationDialog,
  openDeleteConfirmationDialog,
  openProductDialog,
  searchProducts,
} from './products.helper';

@Injectable()
export class ProductsComponentEffect {
  searchProducts$ = createEffect(() =>
    this.actions$.pipe(
      ofType(searchProducts),
      map(({ payload }) => getProducts({ payload }))
    )
  );

  openProductDialog$ = createEffect(
    () =>
      this.actions$.pipe(
        ofType(openProductDialog),
        withLatestFrom(this.store.pipe(select(selectProductEntities))),
        map(
          ([
            {
              payload: { productId },
            },
            entities,
          ]) => entities[productId] ?? null
        ),
        exhaustMap((data) =>
          this.matDialog.open(ProductFormComponent, { data }).afterClosed()
        )
      ),
    { dispatch: false }
  );

  openDeleteConfirmationDialog$ = createEffect(() =>
    this.actions$.pipe(
      ofType(openDeleteConfirmationDialog),
      exhaustMap(({ payload: { productId } }) =>
        this.matDialog
          .open(ConfirmationDialogComponent)
          .afterClosed()
          .pipe(
            map((accept) =>
              accept
                ? acceptDeleteConfirmationDialog({ payload: { productId } })
                : cancelDeleteConfirmationDialog()
            )
          )
      )
    )
  );

  acceptDeleteConfirmationDialog$ = createEffect(() =>
    this.actions$.pipe(
      ofType(acceptDeleteConfirmationDialog),
      map(({ payload }) => deleteProduct({ payload }))
    )
  );

  upsertProductSuccess$ = createEffect(() =>
    this.actions$.pipe(
      ofType(createProductSuccess, updateProductSuccess),
      map(({ type }) =>
        type === createProductSuccess.type
          ? 'Product created!'
          : 'Product updated!'
      ),
      concatMap((message) => [
        closeDialogs(),
        showMessage({ payload: message }),
      ])
    )
  );

  deleteProductSuccess$ = createEffect(() =>
    this.actions$.pipe(
      ofType(deleteProductSuccess),
      concatMap(() => [
        closeDialogs(),
        showMessage({ payload: `Product deleted!` }),
      ])
    )
  );

  createUpdateDeleteFail$ = createEffect(() =>
    this.actions$.pipe(
      ofType(createProductFail, updateProductFail, deleteProductFail),
      map(() =>
        showMessage({ payload: `There was an error please try again later` })
      )
    )
  );

  constructor(
    private readonly actions$: Actions,
    private readonly store: Store<ProductsPartialState>,
    private readonly matDialog: MatDialog
  ) {}
}
