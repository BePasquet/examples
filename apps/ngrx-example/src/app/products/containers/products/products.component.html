<div class="products-container" *ngIf="state$ | async as state">
  <h1>Products</h1>

  <div class="search-container">
    <app-search-input
      (searchChanges)="searchProducts($event)"
    ></app-search-input>
  </div>

  <div class="table-container">
    <app-products-table
      [products]="state.products"
      [total]="state.total"
      (editProduct)="openProductDialog($event.id)"
      (deleteProduct)="confirmDelete($event.id)"
      (changePage)="changePage($event)"
    ></app-products-table>
  </div>
</div>
