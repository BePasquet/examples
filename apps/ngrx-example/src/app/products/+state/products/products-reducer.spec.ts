import { createAction } from '@ngrx/store';
import { errorMessage, product, products } from '../../../test/test.data';
import {
  createProduct,
  createProductFail,
  createProductSuccess,
  deleteProduct,
  deleteProductFail,
  deleteProductSuccess,
  getProducts,
  getProductsFail,
  getProductsSuccess,
  updateProduct,
  updateProductFail,
  updateProductSuccess,
} from './products.actions';
import { productsInitialState, productsReducer } from './products.reducer';

describe('Products Reducer', () => {
  it('Should return initial state on an unknown action', () => {
    const action = createAction('[Unknown]');
    const state = productsReducer(productsInitialState, action);

    expect(state).toEqual(productsInitialState);
  });

  describe('getProducts', () => {
    it('Should remove all entities, set loading to true, loaded to false and clean the error', () => {
      const successAction = getProductsSuccess({ payload: products });

      const populatedState = productsReducer(
        productsInitialState,
        successAction
      );

      const action = getProducts({ payload: { name: '' } });
      const state = productsReducer(populatedState, action);
      const expected = {
        entities: {},
        error: '',
        ids: [],
        loaded: false,
        loading: true,
      };

      expect(state).toEqual(expected);
    });
  });

  describe('getProductsSuccess', () => {
    it('Should populate the state with products, loading to false, loaded to true, and error to an empty string', () => {
      const action = getProductsSuccess({ payload: products });
      const state = productsReducer(productsInitialState, action);

      expect(state).toMatchSnapshot();
    });
  });

  describe('getProductsFail', () => {
    it('Should set loading to false, loaded to false and populate error', () => {
      const action = getProductsFail({ payload: errorMessage });
      const state = productsReducer(productsInitialState, action);

      expect(state).toMatchSnapshot();
    });
  });

  describe('createProduct', () => {
    it('Should set loading state to true and error to empty string', () => {
      const { id, ...payload } = product;
      const action = createProduct({ payload });
      const state = productsReducer(productsInitialState, action);

      expect(state).toMatchSnapshot();
    });
  });

  describe('createProductSuccess', () => {
    it('Should add a product to entities, set loading to false and error to empty string', () => {
      const action = createProductSuccess({ payload: product });
      const state = productsReducer(productsInitialState, action);

      expect(state).toMatchSnapshot();
    });
  });

  describe('createProductFail', () => {
    it('Should set loading to false and populate error', () => {
      const action = createProductFail({ payload: errorMessage });
      const state = productsReducer(productsInitialState, action);

      expect(state).toMatchSnapshot();
    });
  });

  describe('updateProduct', () => {
    it('Should set loading state to true and error to empty string', () => {
      const action = updateProduct({ payload: product });
      const state = productsReducer(productsInitialState, action);

      expect(state).toMatchSnapshot();
    });
  });

  describe('updateProductSuccess', () => {
    it('Should update a product from the entities, set loading to false, and error to empty string', () => {
      const successAction = getProductsSuccess({ payload: products });

      const populatedState = productsReducer(
        productsInitialState,
        successAction
      );

      const action = updateProductSuccess({
        payload: { ...product, name: 'Pre Workout' },
      });

      const state = productsReducer(populatedState, action);

      expect(state).toMatchSnapshot();
    });
  });

  describe('updateProductFail', () => {
    it('Should set loading to false and populate error', () => {
      const action = updateProductFail({ payload: errorMessage });
      const state = productsReducer(productsInitialState, action);

      expect(state).toMatchSnapshot();
    });
  });

  describe('deleteProduct', () => {
    it('Should set loading state to true and error to empty string', () => {
      const action = deleteProduct({ payload: { productId: product.id } });
      const state = productsReducer(productsInitialState, action);

      expect(state).toMatchSnapshot();
    });
  });

  describe('deleteProductSuccess', () => {
    it('Should remove product from entities, set loading to false and error to empty string', () => {
      const successAction = getProductsSuccess({ payload: products });

      const populatedState = productsReducer(
        productsInitialState,
        successAction
      );

      const action = deleteProductSuccess({
        payload: { productId: product.id },
      });

      const state = productsReducer(populatedState, action);

      expect(state).toMatchSnapshot();
    });
  });

  describe('deleteProductFail', () => {
    it('Should set loading to false and populate error', () => {
      const action = deleteProductFail({ payload: errorMessage });
      const state = productsReducer(productsInitialState, action);

      expect(state).toMatchSnapshot();
    });
  });
});
