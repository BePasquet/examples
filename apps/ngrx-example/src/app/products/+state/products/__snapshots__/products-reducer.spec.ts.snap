// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Products Reducer createProduct Should set loading state to true and error to empty string 1`] = `
Object {
  "entities": Object {},
  "error": "",
  "ids": Array [],
  "loaded": false,
  "loading": true,
}
`;

exports[`Products Reducer createProductFail Should set loading to false and populate error 1`] = `
Object {
  "entities": Object {},
  "error": "ERROR",
  "ids": Array [],
  "loaded": false,
  "loading": false,
}
`;

exports[`Products Reducer createProductSuccess Should add a product to entities, set loading to false and error to empty string 1`] = `
Object {
  "entities": Object {
    "1": Object {
      "active": true,
      "categoryId": "1",
      "description": "lorem ipsum",
      "discount": 0,
      "id": "1",
      "images": Array [],
      "name": "Whey Protein",
      "price": 100,
      "rating": 5,
      "stock": 100,
    },
  },
  "error": "",
  "ids": Array [
    "1",
  ],
  "loaded": false,
  "loading": false,
}
`;

exports[`Products Reducer deleteProduct Should set loading state to true and error to empty string 1`] = `
Object {
  "entities": Object {},
  "error": "",
  "ids": Array [],
  "loaded": false,
  "loading": true,
}
`;

exports[`Products Reducer deleteProductFail Should set loading to false and populate error 1`] = `
Object {
  "entities": Object {},
  "error": "ERROR",
  "ids": Array [],
  "loaded": false,
  "loading": false,
}
`;

exports[`Products Reducer deleteProductSuccess Should remove product from entities, set loading to false and error to empty string 1`] = `
Object {
  "entities": Object {
    "2": Object {
      "active": true,
      "categoryId": "1",
      "description": "lorem ipsum",
      "discount": 0,
      "id": "2",
      "images": Array [],
      "name": "Fat Burner",
      "price": 50,
      "rating": 5,
      "stock": 100,
    },
    "3": Object {
      "active": true,
      "categoryId": "1",
      "description": "lorem ipsum",
      "discount": 0,
      "id": "3",
      "images": Array [],
      "name": "Protein Bar",
      "price": 10,
      "rating": 5,
      "stock": 100,
    },
  },
  "error": "",
  "ids": Array [
    "2",
    "3",
  ],
  "loaded": true,
  "loading": false,
}
`;

exports[`Products Reducer getProductsFail Should set loading to false, loaded to false and populate error 1`] = `
Object {
  "entities": Object {},
  "error": "ERROR",
  "ids": Array [],
  "loaded": false,
  "loading": false,
}
`;

exports[`Products Reducer getProductsSuccess Should populate the state with products, loading to false, loaded to true, and error to an empty string 1`] = `
Object {
  "entities": Object {
    "1": Object {
      "active": true,
      "categoryId": "1",
      "description": "lorem ipsum",
      "discount": 0,
      "id": "1",
      "images": Array [],
      "name": "Whey Protein",
      "price": 100,
      "rating": 5,
      "stock": 100,
    },
    "2": Object {
      "active": true,
      "categoryId": "1",
      "description": "lorem ipsum",
      "discount": 0,
      "id": "2",
      "images": Array [],
      "name": "Fat Burner",
      "price": 50,
      "rating": 5,
      "stock": 100,
    },
    "3": Object {
      "active": true,
      "categoryId": "1",
      "description": "lorem ipsum",
      "discount": 0,
      "id": "3",
      "images": Array [],
      "name": "Protein Bar",
      "price": 10,
      "rating": 5,
      "stock": 100,
    },
  },
  "error": "",
  "ids": Array [
    "1",
    "2",
    "3",
  ],
  "loaded": true,
  "loading": false,
}
`;

exports[`Products Reducer updateProduct Should set loading state to true and error to empty string 1`] = `
Object {
  "entities": Object {},
  "error": "",
  "ids": Array [],
  "loaded": false,
  "loading": true,
}
`;

exports[`Products Reducer updateProductFail Should set loading to false and populate error 1`] = `
Object {
  "entities": Object {},
  "error": "ERROR",
  "ids": Array [],
  "loaded": false,
  "loading": false,
}
`;

exports[`Products Reducer updateProductSuccess Should update a product from the entities, set loading to false, and error to empty string 1`] = `
Object {
  "entities": Object {
    "1": Object {
      "active": true,
      "categoryId": "1",
      "description": "lorem ipsum",
      "discount": 0,
      "id": "1",
      "images": Array [],
      "name": "Pre Workout",
      "price": 100,
      "rating": 5,
      "stock": 100,
    },
    "2": Object {
      "active": true,
      "categoryId": "1",
      "description": "lorem ipsum",
      "discount": 0,
      "id": "2",
      "images": Array [],
      "name": "Fat Burner",
      "price": 50,
      "rating": 5,
      "stock": 100,
    },
    "3": Object {
      "active": true,
      "categoryId": "1",
      "description": "lorem ipsum",
      "discount": 0,
      "id": "3",
      "images": Array [],
      "name": "Protein Bar",
      "price": 10,
      "rating": 5,
      "stock": 100,
    },
  },
  "error": "",
  "ids": Array [
    "1",
    "2",
    "3",
  ],
  "loaded": true,
  "loading": false,
}
`;
